src/domain/flow-system/src/entities/shared/batching_rule.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Duration, Utc};
+use thiserror::Error;
+
+use crate::{FlowResult, FlowTrigger};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BatchingRule {
+    min_records_to_await: u64,
+    max_batching_interval: Duration,
+}
+
+impl BatchingRule {
+    const MAX_BATCHING_INTERVAL_HOURS: i64 = 24;
+
+    pub fn new_checked(
+        min_records_to_await: u64,
+        max_batching_interval: Duration,
+    ) -> Result<Self, BatchingRuleValidationError> {
+        if min_records_to_await == 0 {
+            return Err(BatchingRuleValidationError::MinRecordsToAwaitNotPositive);
+        }
+
+        let max_possible_interval = Duration::hours(Self::MAX_BATCHING_INTERVAL_HOURS);
+        if max_batching_interval > max_possible_interval {
+            return Err(BatchingRuleValidationError::MaxIntervalAboveLimit);
+        }
+
+        Ok(Self {
+            min_records_to_await,
+            max_batching_interval,
+        })
+    }
+
+    #[inline]
+    pub fn min_records_to_await(&self) -> u64 {
+        self.min_records_to_await
+    }
+
+    #[inline]
+    pub fn max_batching_interval(&self) -> &Duration {
+        &self.max_batching_interval
+    }
+
+    pub fn evaluate(
+        &self,
+        flow_start_time: DateTime<Utc>,
+        evaluation_time: DateTime<Utc>,
+        triggers: &[FlowTrigger],
+    ) -> BatchingRuleEvaluation {
+        // TODO: it's likely assumed the accumulation is per each input separately, but
+        // for now count overall number
+        let mut accumulated_records_count = 0;
+        let mut watermark_modified = false;
+
+        // Scan each accumulated trigger to decide
+        for trigger in triggers {
+            if let FlowTrigger::InputDatasetFlow(trigger) = trigger {
+                match &trigger.flow_result {
+                    FlowResult::Empty => {}
+                    FlowResult::DatasetUpdate(update) => {
+                        accumulated_records_count += update.num_records;
```

Comment:
If we have datasets:
- A (root)
- B, C, D (derivative) all using A as an input

If A has 1,000 updates, will we have 3,000 triggers (per each downstream dataset) saved in our event store?

I thought that an update of an input will write a trigger once, and all subsequent updates to the same input will not produce a new trigger. I.e. maximum amount of `FlowTrigger::InputDatasetFlow` in our database is capped at the number of inputs used by all derivative datasets.

But here, from how you are summing up `num_records`, it looks like for every update a new trigger will be written. I don't suppose you are updating `num_records` of previous events.

If that's indeed the case this seems highly problematic for high-frequency root datasets that become very popular and are used as inputs of thousands of derivative datasets.

Reply:
I've implemented reduction function for flow triggers that accumulates results of identical triggering inputs, however we will still have large amounts of ES-events in each derived flow, and that is something harder to solve.

Perhaps, any high frequency dataset:
- need a buffering at `PushSource` level anyway
- will anyway be a subject to trottling (currently 1 minute)

Reply:
Closing as it seems this problem was already solved